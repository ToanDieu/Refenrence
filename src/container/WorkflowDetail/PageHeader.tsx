import React, { memo } from "react";
import { connect } from "react-redux";
import { compose } from "redux";
import { createStructuredSelector } from "reselect";
import styled from "@emotion/styled";
import { TranslateFunction } from "react-localize-redux";

import { ProtectedScopedComponent } from "@/components/HocComponent";
import Dropdown from "@/components/units/Dropdown";
import { WorkflowMedia } from "@/resources/workflows";
import { ActionMedia } from "@/resources/actions";
import AutoSaveMessage from "@/components/blocks/AutoSaveMessage";
import { makeSelectTranslate } from "@/i18n/selectors";
import ActiveForm from "@/container/ActivateForm";
import * as actionCreators from "./actions";
import { makeSelectLastSave } from "./selectors";

const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
  margin-bottom: 60px;
`;

const LeftSide = styled.div`
  color: ${props => props.theme.primary};
`;

const Heading = styled.span`
  font-size: 22px;
  margin-right: 10px;
  text-transform: uppercase;
`;

const InlineDropdown = styled(Dropdown)`
  display: inline-block;
`;

interface Props {
  baseID: number;
  workflow: WorkflowMedia;
  actions: ActionMedia[];
  lastSave: number;
  setEditMode: (mode: boolean) => void;
  translate: TranslateFunction;
}

export function PageHeader({
  baseID,
  workflow,
  actions,
  lastSave,
  setEditMode,
  translate
}: Props) {
  return (
    <Wrapper>
      <LeftSide>
        <Heading>{workflow.name}</Heading>
        <AutoSaveMessage
          lastSave={lastSave}
          text={translate("autoSaveText") as string}
          savingText={translate("savingText") as string}
        />
      </LeftSide>
      <div>
        {actions.length > 0 && (
          <React.Fragment>
            <ProtectedScopedComponent
              scopes={["put:my-org-type-base:workflow"]}
            >
              <ActiveForm baseID={baseID} workflow={workflow} />
            </ProtectedScopedComponent>
            <InlineDropdown
              initial="edit"
              items={[
                {
                  id: "edit",
                  name: "Edit Mode"
                },
                {
                  id: "view",
                  name: "View Mode"
                }
              ]}
              iconName="chevron-down"
              onChange={item => {
                if (item && item.id === "edit") {
                  setEditMode(true);
                  return;
                }
                setEditMode(false);
              }}
            />
          </React.Fragment>
        )}
      </div>
    </Wrapper>
  );
}

const mapStateToProps = createStructuredSelector<{}, {}>({
  lastSave: makeSelectLastSave(),
  translate: makeSelectTranslate()
});

const withConnect = connect(
  mapStateToProps,
  { ...actionCreators }
);

export default compose(
  withConnect,
  memo
)(PageHeader) as any;
