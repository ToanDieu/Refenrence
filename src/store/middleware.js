import { matchPath } from "react-router-dom";
import { CHANGE_PAGE, NOTIFY_BUTTON_CLICK } from "../constants/actionTypes";

const changeToBasePage = id => ({
  type: CHANGE_PAGE,
  payload: {
    pageType: "base",
    detail: { id }
  }
});

export const trackPageNavigate = ({
  dispatch
  // getState
}) => next => action => {
  if (action.type === "@@router/LOCATION_CHANGE") {
    console.log("trackViews__action: ", action);
    const pathName = action.payload.location.pathname;

    const isEnteringBaseCases = matchPath(pathName, {
      path: "/bases/:baseID/cases"
    });

    const isEnteringBaseDetail = matchPath(pathName, {
      path: "/bases/:baseID"
    });

    // console.log("isEnteringBaseDetail", isEnteringBaseDetail, action);

    const isEnteringBaseWorkflow = matchPath(pathName, {
      path: "/bases/:baseID/workflows/*/actions"
    });

    const isEnteringNotification = matchPath(pathName, {
      path: "/notifications"
    });

    if (isEnteringNotification) {
      dispatch({
        type: NOTIFY_BUTTON_CLICK
      });
    }

    if (isEnteringBaseCases) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseCases.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }

    if (isEnteringBaseDetail) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseDetail.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }

    if (isEnteringBaseWorkflow) {
      // console.log("trackViews__matchPath: ", isEnteringBaseCases.params.baseID);
      dispatch(changeToBasePage(isEnteringBaseWorkflow.params.baseID));
      // console.log(getState().pageDetail.current);
      return next(action);
    }
  }

  return next(action);
};
