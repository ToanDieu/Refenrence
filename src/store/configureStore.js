import { createStore, compose, applyMiddleware } from "redux";
import reduxImmutableStateInvariant from "redux-immutable-state-invariant";
import thunk from "redux-thunk";
import { routerMiddleware } from "connected-react-router";

import history from "@/utils/history";
import createReducer from "../reducers";
import { trackPageNavigate } from "./middleware";

function configureStoreProd(initialState) {
  const reactRouterMiddleware = routerMiddleware(history);
  const middlewares = [thunk, reactRouterMiddleware, trackPageNavigate];

  const store = createStore(
    createReducer(),
    initialState,
    compose(applyMiddleware(...middlewares))
  );

  // Extensions
  store.injectedReducers = {}; // Reducer registry

  return store;
}

function configureStoreDev(initialState) {
  const reactRouterMiddleware = routerMiddleware(history);
  const middlewares = [
    reduxImmutableStateInvariant(),
    thunk,
    reactRouterMiddleware,
    trackPageNavigate
  ];

  /* eslint-disable no-underscore-dangle */
  const composeEnhancers =
    window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose; // add support for Redux dev tools
  const store = createStore(
    createReducer(),
    initialState,
    composeEnhancers(applyMiddleware(...middlewares))
  );

  // Extensions
  store.injectedReducers = {}; // Reducer registry

  if (module.hot) {
    // Enable Webpack hot module replacement for reducers
    module.hot.accept("../reducers", () => {
      const nextReducer = require("../reducers").default; // eslint-disable-line global-require
      store.replaceReducer(nextReducer);
    });
  }

  return store;
}

const store =
  process.env.NODE_ENV === "production"
    ? configureStoreProd()
    : configureStoreDev();

export { history, store };
