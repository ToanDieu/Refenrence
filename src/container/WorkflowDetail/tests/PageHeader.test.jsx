import React from "react";
import { render, fireEvent } from "react-testing-library";
import { Provider } from "react-redux";
import { initialize, addTranslationForLanguage } from "react-localize-redux";

import languages from "@/i18n/languages.json";
import english from "@/i18n/en.json";
import { store as appStore } from "@/store/configureStore";
import { PageHeader } from "../PageHeader";

describe("<PageHeader />", () => {
  let store;
  const translate = jest.fn();

  beforeAll(() => {
    store = appStore;
    const defaultLanguage = "en";
    store.dispatch(initialize(languages, { defaultLanguage }));
    store.dispatch(addTranslationForLanguage(english, "en"));
  });

  it("should render the header without active button & the mode dropdown", () => {
    const workflow = { id: 1, name: "My test workflow" };
    const actions = [];
    const { container } = render(
      <PageHeader translate={translate} workflow={workflow} actions={actions} />
    );
    expect(container).toMatchSnapshot();
  });

  it("should render the header within active button & the mode dropdown", () => {
    const workflow = { id: 1, name: "My test workflow" };
    const actions = [{ id: 1 }, { id: 2 }];
    const { container } = render(
      <Provider store={store}>
        <PageHeader
          translate={translate}
          workflow={workflow}
          actions={actions}
        />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it("should disable active button if workflow is activated", () => {
    const workflow = { id: 1, name: "My test workflow", active: true };
    const actions = [{ id: 1 }, { id: 2 }];
    const { container } = render(
      <Provider store={store}>
        <PageHeader
          translate={translate}
          workflow={workflow}
          actions={actions}
        />
      </Provider>
    );
    expect(container.querySelector("button")).toMatchSnapshot();
  });

  describe("should toggle edit/view mode", () => {
    it("should toggle mode to edit when click dropdown item", () => {
      const toggleSpy = jest.fn();
      const workflow = { id: 1, name: "My test workflow" };
      const actions = [{ id: 1 }, { id: 2 }];
      const { getAllByText } = render(
        <Provider store={store}>
          <PageHeader
            translate={translate}
            workflow={workflow}
            actions={actions}
            setEditMode={toggleSpy}
          />
        </Provider>
      );
      fireEvent.click(getAllByText("Edit Mode")[0]);
      expect(toggleSpy).toBeCalledWith(true);
    });

    it("should toggle mode to view when click dropdown item", () => {
      const toggleSpy = jest.fn();
      const workflow = { id: 1, name: "My test workflow" };
      const actions = [{ id: 1 }, { id: 2 }];
      const { getAllByText } = render(
        <Provider store={store}>
          <PageHeader
            translate={translate}
            workflow={workflow}
            actions={actions}
            setEditMode={toggleSpy}
          />
        </Provider>
      );
      fireEvent.click(getAllByText("View Mode")[0]);
      expect(toggleSpy).toBeCalledWith(false);
    });
  });
});
