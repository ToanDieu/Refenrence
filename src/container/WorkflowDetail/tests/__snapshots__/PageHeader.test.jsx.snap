// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PageHeader /> should disable active button if workflow is activated 1`] = `null`;

exports[`<PageHeader /> should render the header within active button & the mode dropdown 1`] = `
.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 60px;
}

.emotion-0 {
  font-size: 22px;
  margin-right: 10px;
  text-transform: uppercase;
}

.emotion-2 {
  display: inline-block;
}

<div>
  <div
    class="emotion-3"
  >
    <div
      class="emotion-1"
    >
      <span
        class="emotion-0"
      >
        My test workflow
      </span>
      <span />
    </div>
    <div>
      <div
        class="emotion-2"
        role="presentation"
      >
        <div
          css="[object Object]"
        >
          <div
            role="presentation"
          >
            Edit Mode
          </div>
          <div
            role="presentation"
          >
            View Mode
          </div>
        </div>
        <span>
          Edit Mode
          Â 
          <div
            css="[object Object]"
          >
            <span
              class="icon-chevron-down"
            />
          </div>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`<PageHeader /> should render the header without active button & the mode dropdown 1`] = `
.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 60px;
}

.emotion-0 {
  font-size: 22px;
  margin-right: 10px;
  text-transform: uppercase;
}

<div>
  <div
    class="emotion-2"
  >
    <div
      class="emotion-1"
    >
      <span
        class="emotion-0"
      >
        My test workflow
      </span>
      <span />
    </div>
    <div />
  </div>
</div>
`;
